airflow/utils/sla.py

```diff
@@ -0,0 +1,371 @@
+import airflow.models
+from airflow.utils import asciiart
+from airflow.utils.db import provide_session
+from airflow.utils.email import send_email
+from airflow.utils.log.logging_mixin import LoggingMixin
+from airflow.utils.state import State
+
+log = LoggingMixin().log
+
+
+def yield_unscheduled_runs(dag, last_scheduled_run, ts):
+    """
+    Yield new DagRuns that haven't been created yet.
+    """
+
+    # TODO: A lot of this logic is duplicated from the scheduler. It would
```

Comment:
Thank you for mentioning this. Perhaps this can also be done in this PR.

Reply:
I think that the scheduler code should be revamped but I would be pretty hesitant to touch the scheduler code in this PR. It's a much higher reliability part of the system; if the SLA results diverge from the scheduler results, that is bad, but it is much worse if an SLA improvement inadvertently changes how scheduling works.